server/src/main/java/org/elasticsearch/action/admin/indices/retention/TransportShardSyncRetentionLeasesAction.java

```diff
@@ -0,0 +1,81 @@
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package org.elasticsearch.action.admin.indices.retention;
+
+import java.io.IOException;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.replication.ReplicationResponse;
+import org.elasticsearch.action.support.replication.TransportReplicationAction;
+import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+public class TransportShardSyncRetentionLeasesAction extends
+    TransportReplicationAction<ShardSyncRetentionLeasesRequest, ShardSyncRetentionLeasesRequest, ReplicationResponse> {
+
+    public static final String NAME = SyncRetentionLeasesAction.NAME + "[s]";
+    public static final ActionType<ReplicationResponse> TYPE = new ActionType<>(NAME);
+
+    @Inject
+    public TransportShardSyncRetentionLeasesAction(
+        Settings settings,
+        TransportService transportService,
+        ClusterService clusterService,
+        IndicesService indicesService,
+        ThreadPool threadPool,
+        ShardStateAction shardStateAction) {
+        super(settings, NAME, transportService, clusterService, indicesService, threadPool, shardStateAction, ShardSyncRetentionLeasesRequest::new, ShardSyncRetentionLeasesRequest::new, ThreadPool.Names.FORCE_MERGE);
+    }
+
+    @Override
+    protected ReplicationResponse newResponseInstance(StreamInput in) throws IOException {
+        return new ReplicationResponse(in);
+    }
+
+    @Override
+    protected void shardOperationOnPrimary(ShardSyncRetentionLeasesRequest shardRequest, IndexShard primary, ActionListener<PrimaryResult<ShardSyncRetentionLeasesRequest, ReplicationResponse>> listener) {
+        primary.syncRetentionLeases(new ActionListener<>() {
+            @Override
+            public void onResponse(ReplicationResponse replicationResponse) {
+                listener.accept(new PrimaryResult<>(shardRequest, replicationResponse), null);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    @Override
+    protected ReplicaResult shardOperationOnReplica(ShardSyncRetentionLeasesRequest shardRequest, IndexShard replica) throws Exception {
+        return null;
```

Comment:
Won't this result in an NPE on the replica? Anyhow, can we instead somehow avoid sending the request to the replica at all instead to avoid useless network traffic?

Reply:
The tricky part here is handling asynchronicity.  We have two broadcast handler abstract base classes:
* `BroadcastByNodeTransportAction` - synchronous, but you can filter which shards the request is routed to
* `BroadcastReplicationAction` - async, but sends to both primary and replica
Looking at it, I think the simplest way forward is to add asynchronous handling to `BroadcastByNodeTransportAction` and then use that to handle the retention lease sync.  I'll open a separate PR.

Reply:
It it an option to do something like in TransportKillNodeAction ?

Reply:
Yeah I tried doing something like that, but there's lots of useful infrastructure around picking out specific shards in `BBNTA` that we should be re-using.  I opened https://github.com/crate/crate/pull/16923 to allow subclasses to respond with shard-level results using an ActionListener.